clib/cArray.ml

```diff
@@ -451,6 +452,11 @@ let fold_left2_map f e v1 v2 =
   let v' = map2 (fun x1 x2 -> let (e,y) = f !e' x1 x2 in e' := e; y) v1 v2 in
   (!e',v')
 
+let fold_left2_map_i f e v1 v2 =
+  let e' = ref e in
+  let v' = map2_i (fun idx x1 x2 -> let (e,y) = f idx !e' x1 x2 in e' := e; y) v1 v2 in
```

Comment:
Is this more efficient with a ref?

Reply:
I just copied the code from the non `_i` version. I guess that's a question for experts in low-level OCaml code.

Reply:
Well, given that you just observed non-trivial flambda speedups by removing writes to references, it makes sense to wonder whether avoiding them in those library functions (which sounds quite easy) would again improve performances.

(In general writes to reference can slow things down when they hit the write barrier, unless their static type guarantees that the references don't contain pointers (`int ref` for example). I don't remember offhand if short-lived references are also affected, or only long-lived references. In the context of this PR, it seems that what you observe is rather than flambda is better at optimizing tuple-passing than reference-read-writing code. I think this second effect would not be observed here, unless both `fold_left2_map_i` and `map2_i` were inlined at the point of use -- which you may be able to tell by looking at flambda inliner's reports, but I cannot guess from my chair.)

Reply:
> Well, given that you just observed non-trivial flambda speedups by removing writes to references, it makes sense to wonder whether avoiding them in those library functions (which sounds quite easy) would again improve performances.

IIRC the main reason to use imperative style in these functions may be due to lack of tail-call, but again I am not an expert, so I'll let others handle this.